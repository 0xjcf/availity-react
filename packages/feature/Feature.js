import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BlockUi from 'react-block-ui';
import 'react-block-ui/style.css';
import isFeatureEnabled from './isFeatureEnabled';

class Feature extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
    };
  }

  async checkFeatures() {
    const { loading } = this.state;
    const { features } = this.props;

    if (!loading) this.setState({ loading: true });

    const enabled = await isFeatureEnabled(features);

    this.setState({ enabled, loading: false });
  }

  componentDidMount() {
    this.checkFeatures();
  }

  componentDidUpdate(prevProps) {
    const { features } = this.props;

    if (prevProps.features !== features) {
      if (
        typeof prevProps.features === 'string' ||
        typeof features === 'string'
      ) {
        this.checkFeatures();
      } else {
        const prevFeatures = [].concat(...prevProps.features);
        const currentFeatures = [].concat(...features);
        if (
          prevFeatures.length !== currentFeatures.length ||
          prevFeatures.join() !== currentFeatures.join()
        ) {
          this.checkFeatures();
        }
      }
    }
  }

  render() {
    const { loader, children, whenDisabled, negate } = this.props;
    const { loading, enabled } = this.state;

    if (loading) {
      if (loader) return loader === true ? <BlockUi blocking /> : loader;
      return null;
    }
    const showChildren = enabled ^ negate; // eslint-disable-line no-bitwise
    if (showChildren) {
      return children;
    }

    return whenDisabled;
  }
}

Feature.propTypes = {
  features: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.string,
      ])
    ),
    PropTypes.string,
  ]).isRequired,
  loader: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]),
  whenDisabled: PropTypes.node,
  children: PropTypes.node,
  negate: PropTypes.bool,
};

Feature.defaultProps = {
  whenDisabled: null,
  children: null,
};

export default Feature;
