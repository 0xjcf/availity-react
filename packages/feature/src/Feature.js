import React, { useState } from 'react';
import PropTypes from 'prop-types';
import BlockUi from 'react-block-ui';
import 'react-block-ui/style.css';
import { useEffectAsync } from '@availity/hooks';
import isFeatureEnabled from './isFeatureEnabled';

const Feature = ({ features, loader, whenDisabled, children, negate }) => {
  const [loading, setLoading] = useState(false);
  const [enabled, setEnabled] = useState(null);

  useEffectAsync(async () => {
    if (!loading) setLoading(true);

    const _enabled = await isFeatureEnabled(features);

    setEnabled(_enabled);

    setLoading(false);
  }, [features]);

  if (loading) {
    if (loader) return loader === true ? <BlockUi blocking /> : loader;
    return null;
  }
  const showChildren = enabled ^ negate; // eslint-disable-line no-bitwise
  if (showChildren) {
    return children;
  }

  return whenDisabled;
};

Feature.propTypes = {
  features: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string])),
    PropTypes.string,
  ]).isRequired,
  loader: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]),
  whenDisabled: PropTypes.node,
  children: PropTypes.node,
  negate: PropTypes.bool,
};

Feature.defaultProps = {
  whenDisabled: null,
  children: null,
};

export default Feature;
