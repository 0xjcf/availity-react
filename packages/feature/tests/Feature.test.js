import React from 'react';
import renderer from 'react-test-renderer';
import Feature from '..';

describe('Feature', () => {
  test('should render', () => {
    const component = renderer.create(<Feature features="1234" loader />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
  test('should render with single permission', () => {
    const component = renderer.create(
      <Feature
        features="1234"
        whenDisabled="You do not have permission to see this"
      />
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
  test('should render with array of features', () => {
    const component = renderer.create(
      <Feature
        features={['1234', '2345', ['3456', '4567']]}
        whenDisabled="You do not have permission to see this"
      />
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
  test('should render negate features', () => {
    const component = renderer.create(
      <Feature
        features="1234"
        negate
        whenDisabled="You do not have permission to see this"
      >
        You can see this
      </Feature>
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
