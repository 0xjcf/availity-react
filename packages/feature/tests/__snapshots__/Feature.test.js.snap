// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature should render 1`] = `
<div
  aria-busy={true}
  className="block-ui undefined"
>
  <div
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    tabIndex="0"
  />
  <div
    className="block-ui-container"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    tabIndex="0"
  >
    <div
      className="block-ui-overlay"
    />
    <div
      className="block-ui-message-container"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <div
        className="block-ui-message"
      >
        <div
          className="loading-indicator"
        >
          <span
            className="loading-bullet"
          >
            •
          </span>
           
          <span
            className="loading-bullet"
          >
            •
          </span>
           
          <span
            className="loading-bullet"
          >
            •
          </span>
        </div>
      </div>
    </div>
  </div>
  <span />
</div>
`;

exports[`Feature should render negate features 1`] = `null`;

exports[`Feature should render with array of features 1`] = `null`;

exports[`Feature should render with single permission 1`] = `null`;
